File: lib/ptabler/software/src/ptabler/steps/sort.py
import polars as pl
import msgspec
from typing import List, Optional

from .base import PStep, StepContext
from ..expression import AnyExpression


class SortDirective(msgspec.Struct, frozen=True, rename="camel"):
    """
    Defines a single sort instruction, specifying the expression, sort order,
    and null handling strategy.
    Corresponds to the SortDirective in TypeScript.
    """
    value: AnyExpression
    descending: Optional[bool] = None  # Defaults to False (ascending)
    nulls_last: Optional[bool] = None # Defaults to Polars default (nulls smallest)


class Sort(PStep, tag="sort"):
    """
    PStep to sort a table based on one or more column directives.
    The operation reads from an input table and writes the sorted result
    to an output table.
    
    Sorting is always stable (maintains relative order of equivalent rows)
    for consistent, reproducible results.
    
    Corresponds to the SortStep defined in the TypeScript type definitions.
    """
    input_table: str
    output_table: str
    by: List[SortDirective]

    def execute(self, ctx: StepContext):
        """
        Executes the sort step.

        Args:
            ctx: StepContext containing methods to manage the table space.

        Raises:
            ValueError: If the specified input_table is not found in the tablespace,
                        or if 'by' directives list is empty.
        """
        if not self.by:
            raise ValueError("The 'by' list of sort directives cannot be empty for the sort step.")

        lf = ctx.get_table(self.input_table)

        sort_by_expressions: List[pl.Expr] = []
        descending_flags: List[bool] = []
        nulls_last_flags: List[bool] = []

        for directive in self.by:
            sort_by_expressions.append(directive.value.to_polars())
            
            current_descending = directive.descending if directive.descending is not None else False
            descending_flags.append(current_descending)

            if directive.nulls_last is not None:
                nulls_last_flags.append(directive.nulls_last)
            else:
                # Polars default: nulls are smallest.
                # If ascending (current_descending=False), nulls first (nulls_last=False).
                # If descending (current_descending=True), nulls last (nulls_last=True).
                # So, if directive.nulls_last is None, it should align with current_descending.
                nulls_last_flags.append(current_descending)
        
        sorted_lf = lf.sort(
            by=sort_by_expressions,
            descending=descending_flags,
            nulls_last=nulls_last_flags,
            maintain_order=True  # Always use stable sorting for consistent results
        )

        ctx.put_table(self.output_table, sorted_lf)

File: lib/ptabler/schema/src/sort.ts
import type { Expression } from './expressions';

/**
 * Defines a single sort instruction, specifying the column, sort order,
 * and null handling strategy.
 */
export interface SortDirective {
  /** The expression to sort by. */
  value: Expression;

  /**
   * Whether to sort in descending order for this column.
   * Defaults to false (ascending).
   */
  descending?: boolean;

  /**
   * Determines how null values are handled in the sort for this column.
   * If true, nulls are placed last.
   * If false, nulls are placed first.
   * If undefined, the default Polars behavior is used (nulls are considered smallest,
   * so they appear first in ascending sorts and last in descending sorts, unless platform defaults differ).
   * This maps to Polars' `nulls_last` parameter.
   */
  nullsLast?: boolean;
}

/**
 * Defines a step that sorts a table based on one or more column directives.
 * The operation reads from an input table and writes the sorted result to an output table.
 */
export interface SortStep {
  /** The type identifier for this step. Must be 'sort'. */
  type: 'sort';

  /** The name of the input table from the tablespace to be sorted. */
  inputTable: string;

  /** The name to be assigned to the newly created sorted table in the tablespace. */
  outputTable: string;

  /**
   * An array of sort directives defining the columns and their respective sort orders.
   * The table is sorted by the first directive in the array, then by the second
   * for any ties, and so on.
   */
  by: SortDirective[];

  /**
   * Optional. If true, maintains the stability of the sort. This means that if
   * two rows are equivalent according to the sorting criteria, their relative order
   * from the input table is preserved in the output.
   * Corresponds to `maintain_order=True` in Polars `DataFrame.sort()`.
   * Defaults to false.
   */
  stable?: boolean;
}
